add tutorial on batching approval and swaps

eshaben opened the pull request:
### Description

This PR adds a new tutorial for batching approval and swaps into a single transaction with the `batchAll()` function of the batch precompile.

I'll get to the images and checklist next week.

### Checklist

- [ ] If this requires translations for the `moonbeam-docs-cn` repo, I have created a ticket for the translations in Jira
- [ ] If pages have been moved around, I have created an additional PR in `moonbeam-mkdocs` to update redirects
- [ ] If pages have been moved around, I have run the `move-pages.py` script to move the pages and update the image paths on the chinese repo
    - [ ] After the script has been run, I have created an additional PR in `moonbeam-docs-cn`
- [x] If images have been added, I have run the `compress-images.py` script to compress the images.
- [ ] If variables (in variables.yml) need to be updated (such as a name change), I have updated the `moonbeam-docs-cn` repo to use the new variables
- [x] If this page requires a disclaimer, I have added one

### Corresponding PRs

Please link to any corresponding PRs here.

### After Translation Requirements

- [ ] Will need to create PR in `moonbeam-docs` repo to remove images
- [ ] Will need to create PR in `moonbeam-docs` repo to remove variables
- [ ] Will need to create PR in `moonbeam-mkdocs` repo to add redirects for Chinese site
- [ ] No additional PRs are required after the translations are done

#### Items to be Updated

Please list any of the items that will need to be added or deleted after the translations are done here.

albertov19 commented on tutorials/batch-approve-swap.md:
```diff
@@ -0,0 +1,310 @@
+
+Next, we're going to create a helper function that will be used to check the balance of DTOK tokens the DEX and the signer account has. This will be particularly useful to see balance changes after the swaps are complete.
+
+In this function, you'll call the `balanceOf()` function of the `DemoToken` contract, passing in the address of the signer and the DEX, and then print the formatted results in DTOKs to the terminal:
```
I would state that the contract has an ERC-20 interface so you can check the balance using the normal `balanceOf` function

albertov19 commented on tutorials/batch-approve-swap.md:
```diff
@@ -0,0 +1,310 @@
+  // ...
+
+  // Parse the value to swap to Wei
+  const amountDtok = ethers.utils.parseEther( "INSERT-AMOUNT-OF-DEV-TO-SWAP")
```
`INSERT-AMOUNT-OF-DTOK-TO-SWAP` right?

albertov19 reviewed (changes requested):
Looks good! I was expecting you doing this but against the Uniswap V2 demo app we have in Moonbase, but this is good as well. 

If you want, you can create a section at the end where you talk about how the process would look like when implementing this against an Uniswap V2 styled dex. The most important thin is understanding how to build the encoded call data of [this function](https://github.com/PureStake/moonbeam-uniswap/blob/6e3ecf90fecb77bdea84d320fbaae6512f5e0e69/uniswap-contracts-moonbeam/contracts/periphery/UniswapV2Router02.sol#L284). Which is:

- Amount in
- Amount out
- Path array (in this case it would be [DTOK, ERC-20 PRECOMPILE] I think (or ERC-20 address if you want to use `MARS` for example)
- Recipient Address
- Deadline (TIMESTAMP)

Try it against Moonbeam-Swap first to ensure that it works

eshaben commented:
Thanks for reviewing. I didn't realize that you were thinking about uniswap. Figured I would use a simple example, but I incorporated your feedback and added a section on using the batch logic with uniswap. Just used the existing tokens set up on Moonbeam-swap and to relate back to previous example where we were swapping an ERC-20 for the native asset, I decided to swap ERTH for DEV. Tested it out and it works as expected.

jboetticher commented on tutorials/batch-approve-swap.md:
My initial thought was to ax this section and just have the smart contract directly mint to the deployer's contract, since this section feels like setup that should be removed if possible so that the reader can get to the point of the tutorial faster.

```solidity
contract DemoToken is ERC20 {
    constructor(uint256 initialSupply) ERC20("DemoToken", "DTOK") {
        _mint(msg.sender, initialSupply / 2);
        _mint(tx.origin, initialSupply / 2);
    }
}
contract SimpleDex {
    // Made this payable so that liquidity can be added with deployment
    constructor() payable {
        // Mint 1000 DTOK tokens
        token = new DemoToken(1000000000000000000000);
    }
   //...
}
```

But it also looks like one of the main reasons that this step is needed is so that DEV liquidity exists for the contract. This can also be done in the constructor. If you choose not to remove this section, it would be good to also explain that liquidity is being added for the batch swap

eshaben replied:
Yeah I agree that it's probably easiest to just do this in the constructor. Thanks for bringing this up!

jboetticher reviewed (commented):
Looks good 👍 . The only substantial change that I suggested was to consider replacing the first swap part with constructor logic so that the reader can get to the point of the tutorial faster.

jboetticher commented on tutorials/batch-approve-swap.md:
Potentially add a suggested value?

eshaben replied:
I added a suggested value in the comment

jboetticher reviewed (approved):
👍

eshaben commented:
Thanks for the reviews!
